HULKs/hulk
# Center circle detection for camera calibration

## Why? What?

Provides 3 features, could be splitted but makes it harder to test all together.

1. Center Circle detection with a RANSAC implementation.
    - Canny edge detection or Segments used to get points
    - A variant is provided to handle transforming from Pixel (ellipse) to ground (circle) prior to fitting.
2. Integration into calibration measurement provider.
3. Twix panel to view the detections, a parameter in the circle detection cycler `tuning_mode` triggers detection on every circle regardless of `CalibrationCommand`.

This was originally developed on top of work on branches/ PRs (https://github.com/HULKs/hulk/pull/695)

How it works: from RC2023 Bordeaux
![center_circle](https://github.com/HULKs/hulk/assets/6742239/2ec3128e-3cef-49b4-9d8e-77ab59c555ea)

Differences from the line-RANSAC:

1. Uses 3-point circle formula to get a sample circle
2. Early-rejection of candidates -> uses radius, minimm distance between points, etc.
4. Performing the fit only on a sample of the population (if large enough) to make the iterations reasonably fast.

Testing via #1508 

![image](https://github.com/user-attachments/assets/51a6f049-0c44-4c3f-8b5d-4100c2b74123)


Fixes #

## ToDo / Known Issues

 - [x] Avoid fitting to small arcs - bottom image of the above example. (3 points selected randomly should be sufficiently spaced out)
 - ~[ ] Alternatively, construct the circle from a minimum-sized arc (i.e. 100 degree continuous arc with radius R)~
    - This can be more robust as we can check all points fitting in this region to be continuous or at least mostly continous.

## Ideas for Next Iterations (Not This PR)


## How to Test

- Deploy on a Nao or Webots
- Disable head-unstiff (#1159 ) parameter!!
- place the Nao 1.5~m facing and away from.center circle.center.
- enable calibration mode

You can subscribe to:
- `calibration_command` to see the state changes
- `calibration_inner.measurements` additional from calibration controller to see the measurements collected 
- **Output** `last_calibration_corrections`

### Tuning mode (vision parameters)

The parameters are tuned to a decent state already, but if needed it can be tuned further.

#### Parameters

- `tuning_mode` - keeps running the cycler even when no calibration command is recieving from the calibration controller.
- `ransac_circle_minimum_circumference_percentage` helps to reduce false positive.hits on T junctions, etc with very few points.
- `get_edges_from_segments` - uses segments instead of Canny to generate points for RANSAC Overall speed is quite decent but not as good as Canny based approach 
- `gaussian_sigma` controlls the Gaussian blur strength. Best left untouched.
- `canny_*` - controlls canny edge detection. Can be tuned for lighting conditions, but these values should work already.
- Other parameters are already tuned and does not need changing.

#### How-to

- Deploy the branch on a Nao or Webots
- Place the nao 1.5m~ from the center circle while facing it.
- Enter calibration mode (but keep the head unstiff parameter #1159 to  help tuning)

Twix:

- Open an image panel, there is an overlay for calibration center circle (not fully tested with latest Twix changes!!)
- subscribe to:
    - main output `calibration_center_circles` of the given cycler to see raw.output
    - `calibration_center_circle_detection.circles_points_pixel_scores` additional to see the score for each circle.
- open a parameters panel
    - Look for `calibration_center_circle_detection`
    - Enable `tuning_mode` for top or bottom (not both as this would slow things down.
    - if no circles are detected, try to reduce `ransac_circle_minimum_circumference_percentage` to 0.1 or something, it should
